src/RemovePtrToInt.cc

```diff
@@ -0,0 +1,242 @@
+  auto *PN = dyn_cast<PHINode>(IntVal);
+  if (!PN)
+    return false;
+
+  RPTI_LOG(dbgs() << F.getName() << ":\n");
+  RPTI_LOG(I2P->dump());
+  RPTI_LOG(PN->dump());
+
+  if (NewPhis.count(PN) > 0) {
+    IRBuilder<> IRB(I2P);
+    auto *NewPhi = NewPhis[PN];
+    I2P->replaceAllUsesWith(IRB.CreateBitCast(NewPhi, I2P->getType()));
+    MaybeUnusedInsts.insert(I2P);
+
+    RPTI_LOG(dbgs() << "\n!!!! Reused new PHI ~~~~~ \n");
+    return true;
+  }
+
+  if (!llvm::all_of(PN->incoming_values(),
+                    [](Value *IVal) { return isa<PtrToIntInst>(IVal); })) {
+    RPTI_LOG(
+        dbgs() << "Not all incoming values are PtrToInstInsts, skipping PHI\n");
+    return false;
+  }
+
+  // All confirmed to come from PtrToInt, so we can rewrite the PhiNode.
+  IRBuilder<> IRB(PN);
+  auto *NewPN = IRB.CreatePHI(I2P->getType(), PN->getNumIncomingValues(),
+                              PN->getName() + ".rpti");
+
+  for (unsigned i = 0; i < PN->getNumIncomingValues(); ++i) {
+    auto *IBB = PN->getIncomingBlock(i);
+    Value *IVal = PN->getIncomingValue(i);
+
+    if (auto *P2I = dyn_cast<PtrToIntInst>(IVal)) {
+      assert(DT.dominates(P2I, I2P));
+
+      auto *Ptr = P2I->getPointerOperand();
+      assert(Ptr);
+
+      IRB.SetInsertPoint(P2I);
+      NewPN->addIncoming(IRB.CreateBitCast(Ptr, I2P->getType()), IBB);
+      RPTI_LOG(NewPN->dump());
+
+      MaybeUnusedInsts.insert(P2I);
+    }
+    // It's better teach instcombine not to generate loads we see here.
+    //    else if (auto *LI = dyn_cast<LoadInst>(IVal)) {
+    //      assert(DT.dominates(LI, I2P));
+    //
+    //      auto *Loaded = LI->getPointerOperand();
+    //      Loaded->dump();
+    //
+    //      IRB.SetInsertPoint(LI);
+    //
+    //      auto *CastedLoaded = IRB.CreateBitCast(Loaded,
+    //      I2P->getType()->getPointerTo(0), Loaded->getName() + ".rpti"); auto
+    //      *NewLI = IRB.CreateLoad(CastedLoaded, LI->getName() + ".rpti");
+    //
+    //      NewPN->addIncoming(NewLI, IBB);
+    //      NewLI->dump();
+    //      NewPN->dump();
+    //
+    //      MaybeUnusedInsts.insert(LI);
+    //
+    //      if (LI->getName().find("__x.sroa.0.0.") != size_t(-1)) {
+    //        //__asm__("int3");
+    //        //assert(false);
+    //      }
+    //    }
+  }
+
+  NewPhis[PN] = NewPN;
+  MaybeUnusedInsts.insert(PN);
+
+  I2P->replaceAllUsesWith(NewPN);
+  MaybeUnusedInsts.insert(I2P);
+
+  return false;
+}
+
+bool RemovePtrToInt::runOnFunction(Function &F) {
+  if (F.isDeclaration())
+    return false;
+
+  // Skip special functions
+  if (F.getName().startswith("seahorn.") || F.getName().startswith("shadow.") ||
+      F.getName().startswith("verifier."))
+    return false;
+
+  RPTI_LOG(dbgs() << "\n~~~~~~~ Start of RP2I on " << F.getName()
+                  << " ~~~~~ \n");
+
+  bool Changed = false;
+  auto &DL = F.getParent()->getDataLayout();
+  // Would be better to get the DominatorTreeWrapperPass, but the pass manager
```

Comment:
what is the crash? Maybe it is not initialized properly? Which pass manager crashes, `seapp`?